xaitk_saliency/impls/perturb_image/sliding_radial.py

```diff
@@ -0,0 +1,84 @@
+from typing import Any, Dict, Tuple
+import numpy as np
+from xaitk_saliency.interfaces.perturb_image import PerturbImage
+from scipy.ndimage.filters import gaussian_filter
+
+
+class SlidingRadial (PerturbImage):
+    """
+    Produce perturbation matrices generated by sliding a circle occlusion area
+    with a configured radius over the area of an image. Passing a sigma value
+    will apply a Gaussian filter to the mask, blurring it. This results in a
+    smooth transition from full occlusion in the center of the radial to no
+    occlusion at the edge.
+
+    Due to the geometry of sliding windows, if the stride given does not evenly
+    divide the window size along the applicable axis, then the result plane of
+    values when summing the generated masks will not be even.
+
+    Related, if the stride is set to be larger than the circle diameter, the
+    resulting plane of summed values will also not be even, as there be
+    increasingly long valleys of unperturbed space between masked regions.
+
+    :param radius: The radius of the occlusion area.
+    :param sigma: The sigma value for the Gaussian filter applied to mask.
+    :param stride: The striding step for the center of the radial as a tuple
+        with format `(height_step, width_step)`.
+    """
+
+    def __init__(
+        self,
+        radius: int = 50,
+        sigma: int = 0,
+        stride: Tuple[int, int] = (20, 20),
+    ):
+        self.radius: int = int(radius)
+        self.sigma: int = int(sigma)
+        self.stride: Tuple[int, int] = (int(stride[0]), int(stride[1]))
+
+    def perturb(
+        self,
+        ref_image: np.ndarray
+    ) -> np.ndarray:
+        stride_h, stride_w = self.stride
+        img_size = ref_image.shape[:2]
+        img_h, img_w = img_size
+        center_xs = np.arange(0, img_w, stride_w)
+        center_ys = np.arange(0, img_h, stride_h)
+
+        num_masks = len(center_xs) * len(center_ys)
+        masks = np.empty((num_masks, img_h, img_w), dtype='float32')
+        center_xs_m = np.tile(center_xs, len(center_ys))
+        center_ys_m = np.repeat(center_ys, len(center_xs))
+
+        y, x = np.ogrid[:img_h, :img_w]
+
+        for i, (center_x, center_y) in enumerate(zip(center_xs_m, center_ys_m)):
+            mask = np.zeros((img_h, img_w))
+            circle = (x-center_x)*(x-center_x) + (y-center_y) * \
+                (y-center_y) <= self.radius*self.radius
+            mask[circle] = 1
+
+            if self.sigma:
+                mask = gaussian_filter(mask, sigma=self.sigma)
+                mask = mask/mask.max()
+
+            masks[i] = 1 - mask
```

Comment:
In the loop, we are allocating new memory (`np.zeros((img_h, img_w))`) each time to create a layer of the output masks each time on top of the already allocated all-masks matrix. This seems wasteful as we can just work with a literal slice of the `masks` matrix instead for in-place operation. Additionally, if we just initialize `masks` to ones, we can avoid needing the `masks[i] = 1 - mask` line.
```suggestion
        num_masks = len(center_xs) * len(center_ys)
        masks = np.ones((num_masks, img_h, img_w), dtype='float32')
        center_xs_m = np.tile(center_xs, len(center_ys))
        center_ys_m = np.repeat(center_ys, len(center_xs))

        y, x = np.ogrid[:img_h, :img_w]

        for i, (center_x, center_y) in enumerate(zip(center_xs_m, center_ys_m)):
            mask = masks[i]
            circle = (x-center_x)**2 + (y-center_y)**2 <= self.radius**2
            mask[circle] = 0

            if self.sigma:
                mask[:] = gaussian_filter(mask, sigma=self.sigma)
                mask[:] = mask/mask.max()
```

This results in a "different" mask output, but that's due to seeming floating point error. The absolute summed difference between the original and this impl's output is on the order of `1e-5`.

Reply:
Unfortunately, I think this approach generates different results for the blurring case (i.e. self.sigma > 0). Blurring a circular mask of 0's on a background of 1's is actually different than blurring a circular mask of 1's on a background of 0's and then inverting the final result. I think it has to to do with the final normalization.

Reply:
If it needs to be the original form, then that's what it needs to be. My local naive testing, which included blurring, showed no discernible difference in the mask output (absolute sum total difference of ~1e-5), with literally no difference when applying the masks to uint8 scale space.